connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -279,15 +290,54 @@ public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callb
             throw new IllegalStateException("At least one non-null offset store must be provided");
         }
 
-        return primaryStore.set(values, (primaryWriteError, ignored) -> {
+        List<ByteBuffer> partitionsWithTombstoneOffsets = values.entrySet().stream()
+                .filter(offsetEntry -> offsetEntry.getValue() == null)
+                .map(Map.Entry::getKey).collect(Collectors.toList());
+
+        Map<ByteBuffer, ByteBuffer> tombstoneOffsets = new HashMap<>();
+        for (ByteBuffer partition : partitionsWithTombstoneOffsets) {
+            tombstoneOffsets.put(partition, null);
+        }
+        Map<ByteBuffer, ByteBuffer> regularOffsets = values.entrySet().stream()
+                .filter(offsetEntry -> offsetEntry.getValue() != null)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        // If the supplied offsets contain tombstone values, then tombstone offsets are extracted first,
+        // written to the secondary store in a synchronous manner and then to the primary store.
+        // This is because, if a tombstone offset is successfully written to the per-connector offsets topic,
+        // but cannot be written to the global offsets topic, then the global offsets topic will still contain that
+        // source offset, but the per-connector topic will not. Due to the fallback-on-global logic used by the worker,
+        // if a task requests offsets for one of the tombstoned partitions, the worker will provide it with the
+        // offsets present in the global offsets topic, instead of indicating to the task that no offsets can be found.
+        CompletableFuture<Void> offsetWriteFuture = CompletableFuture.completedFuture(null);
+        if (secondaryStore != null && !tombstoneOffsets.isEmpty()) {
+            offsetWriteFuture.thenAccept(v -> {
+                Future<Void> secondaryWriteFuture = secondaryStore.set(tombstoneOffsets, new FutureCallback<>());
+                try {
+                    if (exactlyOnce) {
+                        secondaryWriteFuture.get();
+                    } else {
+                        secondaryWriteFuture.get(offsetFlushTimeoutMs, TimeUnit.MILLISECONDS);
+                    }
+                } catch (ExecutionException e) {
+                    log.error("{} Flush of tombstone(s) offsets to secondary store threw an unexpected exception: ", this, e.getCause());
+                    callback.onCompletion(e.getCause(), null);
+                } catch (Exception e) {
+                    log.error("{} Got Exception when trying to flush tombstone(s) offsets to secondary store", this, e);
+                    callback.onCompletion(e, null);
+                }
+            });
+        }
+        offsetWriteFuture.thenAccept(v -> primaryStore.set(values, new FutureCallback<>((primaryWriteError, ignored) -> {
```

Comment:
We also need to update `offsetWriteFuture` to the return value of `CompletableFuture::thenAccept` (or `completableFuture::thenRun`) here:
```suggestion
        offsetWriteFuture = offsetWriteFuture.thenRun(() -> primaryStore.set(values, new FutureCallback<>((primaryWriteError, ignored) -> {
```

We should probably also add tests that ensure that the future returned from `ConnectorOffsetBackingStore::set` doesn't return until all necessary writes to the primary and secondary stores have completed (with the exception of non-tombstone writes to the secondary store).

Reply:
I have not added this test at this point and I can add it if you think it's still needed with the new approach.